controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,194 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// once is a pointer to the sync.Once type
+var once sync.Once
+
+// funcError contains the error that occurred inside the once.Do() function call
+var funcErr error
+
+// controllerInit is a bool that denotes if the controller is reconciling for the first time
+var controllerInit bool
+
+// NamespaceReconciler reconciles the namespaces starting with openshift-*
+// nolint:revive
+type NamespaceReconciler struct {
+	client.Client
+	ctx    context.Context
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=csiaddons.openshift.io,resources=pods,verbs=list;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch;update;patch
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// i.e,the annotations of the namespace
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	prevLogger := r.Log
+	defer func() { r.Log = prevLogger }()
+	r.Log = r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	r.ctx = ctx
+
+	r.Log.Info("Reconciling Namespace.", "Namespace", klog.KRef(request.Namespace, request.Name))
+
+	controllerInit = false
+	once.Do(func() {
+		nsList := &corev1.NamespaceList{}
+		err := r.List(ctx, nsList)
+		if err != nil {
+			r.Log.Error(err, "Failed to list Namespaces")
+			funcErr = err
+		}
+
+		openshiftNamespaces := []corev1.Namespace{}
+		for _, ns := range nsList.Items {
+			if strings.HasPrefix(ns.Name, "openshift-") {
+				openshiftNamespaces = append(openshiftNamespaces, ns)
+			}
+		}
+		_, funcErr = r.annotateOpenshiftNamespaces(openshiftNamespaces)
+
+		funcErr = nil
+		controllerInit = true
+	})
+
+	if funcErr != nil {
+		return reconcile.Result{}, funcErr
+	}
+
+	updated := false
+	if !controllerInit {
+		ns := &corev1.Namespace{}
+		err := r.Get(ctx, request.NamespacedName, ns)
+		if err != nil {
+			r.Log.Error(err, fmt.Sprintf("Namespace %s does not exist", request.Name))
+			return reconcile.Result{}, err
+		}
+
+		// Annotate namespaces starting with openshift-*, but ignore namespace if
+		// it has annotation "'skipReclaimspaceSchedule':'true'" set
+		nsList := []corev1.Namespace{*ns}
+		updated, err = r.annotateOpenshiftNamespaces(nsList)
+		if err != nil {
+			r.Log.Error(err, "Failed annotating openshift-* namespaces")
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Restart the csi-addons controller pod if any namespace was annotated
+	if updated || controllerInit {
+		podList := &corev1.PodList{}
+		listOpts := []client.ListOption{
+			client.InNamespace(request.NamespacedName.Namespace),
```

Comment:
We can not make use of `request.NamespacedName.Namespace`. Instead get the namespace from `OPERATOR_NAMESPACE` env variable.